### PR TITLE
 Convert Python license classifiers to SPDX expressions

### PR DESCRIPTION
**Pull request by befeleme**
To make it work it was needed to:
- change the Trove - Fedora mappings (remove Callaway expressions, add SPDX)
- delegate the compliance check of Fedora licenses to a source maintained by the Fedora Legal team
- move betamax configuration to conftest.py to achieve a reliable global setting
- rewrite several tests that were blocked by licensing issues of the previously used packages
- regenerate all betamax cassettes

This is a best-effort solution which can't (nor aims to) cover the edge-cases. Unless Python adopts SPDX standard, this will always be approximation.

### CODE REVIEW

File: pyp2spec/license_processor.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    def __init__(self, classifiers):
+        self.classifiers = classifiers
+        self.spdx_identifiers = self.convert_to_spdx()
+
+    def load_trove2fedora_map(self):
+        """Load the file with Python trove classifiers mapped to SPDX identifiers."""
+
+        with open_text("pyp2spec", "classifiers_to_fedora.json") as map_data:
+            return json.load(map_data)
```
Comment (hroncok) [inline pyp2spec/license_processor.py:22]: I don't know how many times a TroveConverter class is actually used when the app runs, but would it make sense to cache the content of this file globally instead of reading it from disk every time this class is used?

File: pyp2spec/license_processor.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        if not self.is_valid_spdx_expression():
+            raise ValueError
+        return self.license_expression
+
+
+def load_fedora_licenses(session):
+    """Load the list of licenses evaluated for Fedora by Fedora Legal team.
+    """
+
+    url = "https://gitlab.com/fedora/legal/fedora-license-data/-/jobs/artifacts/main/raw/fedora-licenses.json?job=json"
```
Comment (hroncok) [inline pyp2spec/license_processor.py:105]: In Fedora, I'd rather load this json from the Fedora package. Is that possible/planned?

File: pyp2spec/license_processor.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        trove2fedora_map = self.load_trove2fedora_map()
+        spdx_identifiers = []
+        for classifier in self.classifiers:
+            try:
+                fedora_identifier = trove2fedora_map[classifier]
+            except KeyError:
+                err_string = """{0}: Such classifier doesn't exist.
+                If you believe that's pyp2spec's error, open an issue at project's GitHub repo."""
+                click.secho(err_string.format(classifier), fg="red")
+                sys.exit(1)
```
Comment (hroncok) [inline pyp2spec/license_processor.py:41]: Style note (and if the rest of the app does it this way, feel free to disregard that here and let's talk about it later). I consider methods and functions calling exit a code smell. Raising an exception (other than SystemExit, e.g. ValueError here) and handling exit from the main app entry point is IMHO much cleaner. I don't know if this is just my view or some generally agreed-upon dogrma :)

**Review by hroncok — COMMENTED** [review]
Not a full review, just a few notes that I've been gathering for couple days.

File: pyp2spec/license_processor.py (hunk 2 above)
Comment (befeleme) [reply pyp2spec/license_processor.py:70]: This feature is added now. First, there's an attempt to read the data from the local file, and only when not found - on the internetz.

File: pyp2spec/license_processor.py (hunk 3 above)
Comment (befeleme) [reply pyp2spec/license_processor.py]: I removed all the `sys.exit`s from license_processor.

**Comment by befeleme** [comment]
Moved to draft, I also want to simplify `is_good_for_fedora` to escape the indentation hell.

File: pyp2spec/pyp2conf.py [hunk 4]
```diff
@@ -14,24 +14,16 @@
...
 
 class PypiPackage:
-    """Process the package data obtained from PyPI."""
-
-    def __init__(self, package, session=None):
-        self.package = package
-        self.package_data = get_package_metadata(package, session=session)
+    """Store and process the package data obtained from PyPI."""
+
+    def __init__(self, package_name, package_metadata=None, session=None):
```
Comment (befeleme) [inline pyp2spec/pyp2conf.py:21]: `self, package_name, *, package_metadata=None, session=None`

File: pyp2spec/pyp2conf.py [hunk 5]
```diff
@@ -123,11 +127,13 @@ def transform_to_spdx(self):
 
         license_keyword = self.package_data["info"]["license"]
         identifiers = license_keyword_to_spdx_identifiers(license_keyword)
+
+        # No more options to detect licenses left, hence explicit fail
         if not identifiers:
             raise ValueError("No license detected")
```
Comment (befeleme) [inline pyp2spec/pyp2conf.py:133]: `NoLicenseDetectedError(ValueError)`

**Comment by befeleme** [comment]
The finding from the latest review were incorporated and all squashed together.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,13 @@
 # Changelog
 
+# Unreleased
+### Changed
+- Convert the Python classifiers to Fedora SPDX expressions (where possible, based on PEP-0639)
```
Comment (hroncok) [inline CHANGELOG.md:5]:
```suggestion
- Convert the Python classifiers to Fedora SPDX expressions (where possible, based on [PEP 639](https://peps.python.org/pep-0639/))
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,13 @@
...
+- When `--fedora-compliant` option is set, a check of the converted expressions is performed
+to determine license is good for Fedora. Data is provided by the [Fedora Legal team](https://gitlab.com/fedora/legal/fedora-license-data/)
```
Comment (hroncok) [inline CHANGELOG.md:7]:
```suggestion
- When `--fedora-compliant` option is set, a check of the converted license expressions is performed
to determine whether the license is allowed in Fedora. Data is provided by the [Fedora Legal team](https://gitlab.com/fedora/legal/fedora-license-data/)
```

File: pyp2spec/pyp2conf.py [hunk 8]
```diff
@@ -34,13 +42,13 @@ def normalize(self, package_name):
 
         return re.sub(r"[-_.]+", "-", package_name).lower()
 
-    def get_package_metadata(self, session):
-        pkg_index = f"https://pypi.org/pypi/{self.package}/json"
+    def get_package_metadata(self, session=None):
```
Comment (hroncok) [inline pyp2spec/pyp2conf.py:45]:
```suggestion
    def get_package_metadata(self, *, session=None):
```

File: pyp2spec/pyp2conf.py [hunk 9]
```diff
@@ -92,71 +100,70 @@ def pypi_version_or_macro(self, version):
...
+            if identifiers:
+                expression = " AND ".join(identifiers)
+                return (identifiers, expression)
+
+        license_keyword = self.package_data["info"]["license"]
+        identifiers = license_keyword_to_spdx_identifiers(license_keyword)
+
+        # No more options to detect licenses left, hence explicit fail
+        if not identifiers:
+            raise NoLicenseDetectedError
```
Comment (hroncok) [inline pyp2spec/pyp2conf.py:136]: I personally dislike raising exception classes (rather than instances), but it's probably not wrong :confused:

File: requirements.txt [hunk 10]
```diff
@@ -1,5 +1,6 @@
 click
 jinja2
+license_expression
```
Comment (hroncok) [inline requirements.txt:3]: This is both the name upstream uses and the PEP 503 normalized name:

```suggestion
license-expression
```

File: setup.cfg [hunk 11]
```diff
@@ -25,6 +25,7 @@ python_requires = >=3.8
 install_requires =
     click
     jinja2
+    license_expression
```
Comment (hroncok) [inline setup.cfg:28]:
```suggestion
    license-expression
```

**Review by hroncok — APPROVED** [review]
Overall looking good. Added a couple opinionated suggestions.

**Comment by befeleme** [comment]
> Overall looking good. Added a couple opinionated suggestions.

Thank you, I kept all of them.
